### PR TITLE
config: remove expand-external-labels flag in release 3.0

### LINKED ISSUE #14486: config: remove `expand-external-labels` feature flag in v3
### Proposal

I believe we should make use of the v3 release to finally remove the `expand-external-labels` feature flag and make the behavior "always on". The impact was anyway very low in practice, but it could have broken users in rare cases. The v3 release is the perfect opportunity to get this technically breaking change out of the feature flag limbo.

### PR DESCRIPTION
**Pull request by jyz0309**
Fixes #14486 
Remove expand-external-labels flag in release-3.0, and make the behavior "always on".
And also remove it in doc.

### CODE REVIEW

**Comment by jyz0309** [comment]
> This behavior should be documented in docs/configuration/configuration.md

This is a command line flag, and I see that `docs/configuration/configuration.md` describes the configuration file, so I've removed the description of this flag in `docs/feature_flags.md`

**Comment by jan--f** [comment]
We have merge `release-3.0` into `main`. The base branch for this PR should be changed to `main` now.

**Comment by jan--f** [comment]
@jyz0309 The changes currently don't compile and the doc addition is still missing. Do you still plan to work on this?

**Comment by jyz0309** [comment]
Sorry that I've been so busy in the past month with my new job...I've added this in doc and fixed the test. PTAL@jan--f

File: config/config.go [hunk 1]
```diff
@@ -85,7 +85,7 @@ func Load(s string, expandExternalLabels bool, logger log.Logger) (*Config, erro
 		return nil, err
 	}
 
-	if !expandExternalLabels {
+	if cfg.GlobalConfig.ExternalLabels.Len() == 0 {
 		return cfg, nil
 	}
```
Comment (jan--f) [inline config/config.go:90]: I think we want to drop this conditional completely.

File: config/config_test.go [hunk 2]
```diff
@@ -2136,38 +2136,34 @@ func TestExpandExternalLabels(t *testing.T) {
 	// Cleanup ant TEST env variable that could exist on the system.
 	os.Setenv("TEST", "")
 
-	c, err := LoadFile("testdata/external_labels.good.yml", false, false, log.NewNopLogger())
-	require.NoError(t, err)
-	testutil.RequireEqual(t, labels.FromStrings("bar", "foo", "baz", "foo${TEST}bar", "foo", "${TEST}", "qux", "foo$${TEST}", "xyz", "foo$$bar"), c.GlobalConfig.ExternalLabels)
```
Comment (jan--f) [inline config/config_test.go:2141]: Why drop this test?

File: docs/configuration/configuration.md [hunk 3]
```diff
@@ -76,7 +76,8 @@ global:
   [ rule_query_offset: <duration> | default = 0s ]
 
   # The labels to add to any time series or alerts when communicating with
-  # external systems (federation, remote storage, Alertmanager).
+  # external systems (federation, remote storage, Alertmanager). 
+  # Note that external_labels are always enabled in release-3.0.
```
Comment (jan--f) [inline docs/configuration/configuration.md:80]: External labels are always added. what we need here is a short description of how env variables are expanded. see `docs/feature_flags.md`.

**Review by jan--f — CHANGES_REQUESTED** [review]
Thanks @jyz0309 !
I think this doesn't quite do the right thing yet. It does though when we drop the conditional in `config/config.go`. Please read https://github.com/prometheus/prometheus/blob/main/docs/feature_flags.md#expand-environment-variables-in-external-labels
also I think this is missing two things: Remove the feature documentation from `docs/feature_flags.md` and add the details about the env var expansion to `docs/configuration/configuration.md`.

File: docs/configuration/configuration.md (hunk 3 above)
Comment (jyz0309) [reply docs/configuration/configuration.md:80]: How about add a example just like `# Example: exported_instance:127.0.0.1`?

File: config/config.go (hunk 1 above)
Comment (jyz0309) [reply config/config.go:90]: But although external labels have been enabled by default, can still be left empty in the configuration file, and I'm worried that removing this conditional completely will cause some tests to failed.

Comment (jan--f) [reply config/config.go:90]: [Below in line 93](https://github.com/prometheus/prometheus/pull/14657/files#diff-fe44f09c4d5977b5f5eaea29170b6a0748819c9d02271746a20d81a5f3efca17R93) this code iterates over the configured external labels. So the check `cfg.GlobalConfig.ExternalLabels.Len() == 0` is redundant.

File: docs/configuration/configuration.md (hunk 3 above)
Comment (jan--f) [reply docs/configuration/configuration.md]: No this is not sufficient. External labels have always been enabled. This change here is about Prometheus expanding environment variable references in the external labels config. This feature used to be behind a feature flag, now this expanding behavior is enabled by default.
So we need to document how this environment variable expansion works. Here we need to add something like
```
Environment variable references `${var}` or `$var` are replaced according to the values of the current environment variables. References to undefined variables are replaced by the empty string.
The `$` character can be escaped by using `$$`.
```

File: config/config_test.go (hunk 2 above)
Comment (jan--f) [reply config/config_test.go:2141]: This test still needs addressing. Please make sure you understand what this change is implementing.

Comment (jyz0309) [reply config/config_test.go:2141]: This test is to verify the parsing when `expand-external-labels` is set to false, but now that `expand-external-labels` is true by default, the result will be the same as other tests in this func, so I think we can just remove it.
if you have a better solution🤔, plz tell me :) thx

Comment (jan--f) [reply config/config_test.go:2141]: Sorry, yes that is what I meant. My original comment meant to read "Why *not* drop this test".
Sorry for the confusion...I agree this test should be dropped.

**Review by jan--f — APPROVED** [review]
Thanks, this looks good to me. Can you please squash your commits?

@roidelapluie are you happy with the `configuration.md` addition?

File: config/config.go [hunk 4]
```diff
@@ -85,10 +85,6 @@ func Load(s string, logger log.Logger) (*Config, error) {
 		return nil, err
 	}
 
-	if cfg.GlobalConfig.ExternalLabels.Len() == 0 {
```
[the remark this answers is not in the document]
Comment (jan--f) [reply config/config.go]: It's not needed.

[the remark this answers is not in the document]
Comment (jan--f) [reply config/config.go]: We call `b := labels.NewScratchBuilder(0)` below, which calls `ScratchBuilder{add: make([]Label, 0, 0)}`. Isn't that effectively a nil slice?

[the remark this answers is not in the document]
Comment (jan--f) [reply config/config.go]: @roidelapluie Would you still prefer to keep this change or can this move ahead as is?

**Comment by jan--f** [comment]
Let's go ahead with this. We can address any optimizations in a follow-up if needed.

**Pull request merged**